### ISSUE TITLE
Eval bug: failed slot restore leaves corrupted K/V data that breaks subsequent inference

### ISSUE THREAD

**Issue by CHIPMUNK-T0T** [ggml-org/llama.cpp#27068]
### Name and Version

version: 0.1.0-dev (build 10398, commit 8e7f22b67)
built with GNU 13.3.0 for Linux x86_64

### Operating systems

Linux

### GGML backends

CUDA
CPU

### Hardware

Intel Core i7 14700F
NVIDIA GeForce RTX 4070 12 GB

### Models

Original observation through Ollama:

* `Qwen3.5 2B Q8_0` (hybrid)

Primary `llama-server` reproduction:

* `Gemma 3 4B IT Q4_K_M`

### Problem description & steps to reproduce

I found this while intentionally corrupting a completed slot snapshot as part of a persistent-cache fault-injection test through Ollama.
I have not observed slot files becoming corrupted during normal save or restore operations.
The corrupted state is correctly rejected, but K/V data written before the validation failure can remain in the cache and affect subsequent inference.
In the reproducing case, later completions become empty, erasing the slot does not recover the process, and restarting `llama-server` does.

I reproduced the same behavior with stock `llama-server`, so this is not specific to Ollama.
`llama-completion` does not exercise this path because its prompt-cache path uses the whole-context state API, while this issue is in the per-sequence slot restore path.

### Steps to reproduce

1. Start `llama-server`:

```bash
llama-server -m gemma-3-4b-it-Q4_K_M.gguf --no-mmproj \
  --slot-save-path /tmp/slots \
  -np 1 -c 4096 -ngl all -fa on --no-webui
```

2. Process the prompt used in this reproduction:

```bash
curl -X POST http://127.0.0.1:8080/completion \
  -H "Content-Type: application/json" \
  -d "$(python3 -c "import json; print(json.dumps({
      'prompt': 'The quick brown fox jumps over the lazy dog near the river bank. ' * 170,
      'n_predict': 1,
      'temperature': 0,
      'cache_prompt': True
  }))")"
```

This produced about 2382 prompt tokens.

Save slot 0:

```bash
curl -X POST "http://127.0.0.1:8080/slots/0?action=save" \
  -H "Content-Type: application/json" \
  -d '{"filename":"good.bin"}'
```

The resulting state file was `170,469,452` bytes.

3. Corrupt 8 MiB starting at 14 MiB:

```bash
cp /tmp/slots/good.bin /tmp/slots/corrupt.bin

dd if=/dev/urandom of=/tmp/slots/corrupt.bin \
  bs=1048576 count=8 seek=14 conv=notrunc
```

These offsets are specific to this saved state file. They caused some K/V data to be written to the backend buffers before validation failed on a later tensor header, at layer 29 in this run.

As controls, a truncated copy and loading the same state with the incompatible `-fa off` layout both failed before entering the K/V data-write loop. Neither affected subsequent inference.

4. Restore the corrupted state:

```bash
curl -X POST "http://127.0.0.1:8080/slots/0?action=restore" \
  -H "Content-Type: application/json" \
  -d '{"filename":"corrupt.bin"}'
```

The restore correctly returns HTTP 400.

5. Send a short completion request:

```bash
curl -sS -X POST http://127.0.0.1:8080/completion \
  -H 'Content-Type: application/json' \
  -d '{"prompt":"Write one short sentence about the sea.","n_predict":32,"temperature":0,"cache_prompt":false}'
```

In the reproducing case, the relevant response fields are:

```text
tokens_predicted: 32
stop_type: "limit"
content: ""
```

6. Erase slot 0, then repeat the same completion request.

The response remains empty and the process remains affected.

7. Restart `llama-server`. Completions work normally again.

### Suspected cause

`state_read_data()` validates and writes K/V data layer by layer.
If validation fails on a later layer, earlier K/V data may already have been written into the backend buffers.

The rollback is asymmetric:

* whole-context restore uses `clear(true)`, which also clears the K/V buffers
* per-sequence restore uses `seq_rm()`, which removes cell metadata but leaves the underlying K/V data

This can leave corrupted K/V data in cells that are subsequently considered free.
Free cells are expected to be excluded by attention masking. In the reproduced path, corrupted f16 data can contain NaNs that are not neutralized by the masking operation.
A C API reproduction also confirmed NaN logits on CPU and showed that deferred buffer writes can be applied after the existing rollback.
PR #26051 made exceptions from `state_read_data()` enter the existing rollback path. For per-sequence restore, that path removes sequence metadata but does not clear K/V data written before the failure.

### Expected behavior

A failed per-sequence restore should leave later inference and other sequences usable. Restoring the previous contents of the target slot is not required.

### Related

* #26744 - stale K/V in freed cells affects attention through a separate Vulkan-specific mechanism
* #9173 - slot restore failure followed by failure in later processing
* #17527 / PR #17982 - fragmented restore failure before K/V data is written
* #23720 / PR #23981 - checkpoint restore failure involving internally generated checkpoint state

AI tools assisted with code investigation and analysis. I independently reproduced the issue and validated the technical claims in this report.

A minimal fix and regression test are working locally, and I can open a PR. 

### First Bad Commit

I do not believe this is a recent regression. The metadata-only per-sequence rollback predates #26051, and this reproduction reaches it through `state_read_data()` returning `false`.

### Relevant log output

```console
state_read_data: mismatched key type (1 != 248383823, layer 29)
llama_state_seq_load_file: error loading sequence state file: failed to restore kv cache
srv    send_error: task id = 5, error: Unable to restore slot: No available space in KV cache or invalid slot save file
```

The mismatched key value comes from the randomly corrupted bytes and therefore changes between runs.
